app/src/main/java/org/mozilla/fenix/home/recentvisits/RecentVisitsFeature.kt

```diff
@@ -0,0 +1,161 @@
+                    HistoryHighlightWeights(MIN_VIEW_TIME_OF_HIGHLIGHT, MIN_FREQUENCY_OF_HIGHLIGHT),
+                    MAX_RESULTS_TOTAL
+                )
+            }
+
+            val recentHistoryGroups = async {
+                historyMetadataStorage.getHistoryMetadataSince(Long.MIN_VALUE)
+                    .filter { it.totalViewTime > 0 && it.key.searchTerm != null }
+                    .groupBy { it.key.searchTerm!! }
+                    .mapValues { group ->
+                        // Within a group, we dedupe entries based on their url so we don't display
+                        // a page multiple times in the same group, and we sum up the total view time
+                        // of deduped entries while making sure to keep the latest updatedAt value.
+                        val metadataInGroup = group.value
+                        val metadataUrlGroups = metadataInGroup.groupBy { metadata -> metadata.key.url }
+                        metadataUrlGroups.map { metadata ->
+                            metadata.value.reduce { acc, elem ->
+                                acc.copy(
+                                    totalViewTime = acc.totalViewTime + elem.totalViewTime,
+                                    updatedAt = max(acc.updatedAt, elem.updatedAt)
+                                )
+                            }
+                        }
+                    }
+                    .toList()
+            }
+
+            updateState(historyHighlights.await(), recentHistoryGroups.await())
+        }
+    }
+
+    @VisibleForTesting
+    internal fun updateState(
+        historyHighlights: List<HistoryHighlight>,
+        historyGroups: List<Pair<String, List<HistoryMetadata>>>
+    ) {
+        homeStore.dispatch(
+            HomeFragmentAction.RecentHistoryChange(
+                getCombinedHistory(historyHighlights, historyGroups)
+            )
+        )
+    }
+
+    @VisibleForTesting
+    internal fun getCombinedHistory(
+        historyHighlights: List<HistoryHighlight>,
+        historyGroups: List<Pair<String, List<HistoryMetadata>>>
+    ): List<RecentlyVisitedItem> {
+        val dedupedHistoryHighlights = dedupeVisits(historyHighlights, historyGroups)
+        val recentHighlights = dedupedHistoryHighlights
+            .map {
+                RecentHistoryHighlight(
+                    title = if (it.title.isNullOrBlank()) it.url else it.title!!,
+                    url = it.url
+                )
+            }
+        val recentGroups = historyGroups
+            .map { (title, data) ->
+                RecentHistoryGroup(
+                    title = title,
+                    historyMetadata = data
+                )
+            }
+            .sortedByDescending { it.lastUpdated() }
+            .take(MAX_RESULTS_TOTAL)
+
+        val totalItemsCount = recentHighlights.size + recentGroups.size
+
+        return if (totalItemsCount <= MAX_RESULTS_TOTAL) {
```

Comment:
We verified that we update the requirements as this is a better solution here! 👍 (but two changes below.)